[V2 MAIN BRANCH] ZNS v2 for Z Chain [WIP]

Whytecrowe opened the pull request.

JamesEarle commented on contracts/price/ZNSCurvePricer.sol:
```diff
@@ -250,19 +293,26 @@ contract ZNSCurvePricer is AAccessControlled, ARegistryWired, UUPSUpgradeable, I
-        require(multiplier <= 10**18, "ZNSCurvePricer: precisionMultiplier cannot be greater than 10^18");
-        priceConfigs[domainHash].precisionMultiplier = multiplier;
-
+        _setPrecisionMultiplier(domainHash, multiplier);
```
If all the setters call to internal functions that both validate and set the variable they should be named the same, `_validateSet[VAR_NAME]` like above with `_validateSetMaxLength(...)`

MichaelKorchagin committed:
Added supports interface for AC. Updated `setAC` function.

MichaelKorchagin committed:
Uncommented line to check CI errors in a different environment.

MichaelKorchagin committed:
Corrected supportsInterface function. Updated call to this. Skipped tests for next pr..

MichaelKorchagin committed:
Added test for common setAC functions for all contracts.

MichaelKorchagin committed:
Splitting tests from one to separate, private ones. Added getInterfaceFunction for each contract which inherit AC.

MichaelKorchagin committed:
SupportsInterface() testing

JamesEarle committed:
add and verify ability to encode and decode struct data correctly

JamesEarle committed:
wip add templates for extra funcs in registry to for additional data

JamesEarle committed:
add storage in registry for pricer and priceConfig, create getters and setter for pair

Whytecrowe committed:
create proto ControlledRegistrar

Whytecrowe committed:
add some TODOs to contracts

Whytecrowe committed:
make getConfig assign deployer zeroVaultAddress if not passed

Whytecrowe committed:
start the test for ControlledRegistrar

Whytecrowe committed:
move domain validation logic to coreRegister , add testsfrom both registrars

JamesEarle committed:
WIP add flow in registrars (mostly) and specifics for fixed pricer as well

JamesEarle committed:
WIP flesh out flow more with registrars and update test helpers, tests for curve pricer passing, updates to fixed pricer tests

Whytecrowe committed:
rework transferFrom logic:

-  in the DomainToken to allow only owner of both: registry and domain token to transfer the token
- add reclaim() function that overrides the transfer and moves token to the registry owner called by Registrar only

Whytecrowe committed:
rework reclaim in RootRegistrar to reclaim token instead of Registry record

Whytecrowe committed:
add some new tests

Whytecrowe committed:
small fixes for tests and getConfig

Whytecrowe committed:
change registerSubdomain() args to struct, pass tokenOwner param through all Registrars to make token mint to another user at registration

Whytecrowe committed:
change revokeDomain() AC from requiring ownership of both to only owner in Registry

Whytecrowe committed:
try testing without using ControlledRegistrar

Whytecrowe committed:
fix test to work under this describe

Whytecrowe committed:
remove ControlledRegistrar contract and remove it from test

Whytecrowe committed:
rework reclaim into registry owner override of token only transfer:

- change reclaimDomainToken to validate AC registry owner only and call the new transferOverride on DomainToken
- rename DomainToken.reclaim() to transferOverride()
- rework DomainToken.isFullyOwned() to isControlled()

Whytecrowe committed:
refactor and group controlled domain test so far

Whytecrowe committed:
rework registerSubdomain() to let operators register on locked parent, but use parent owner as new subdomain owner

Whytecrowe committed:
add and refactor tests to run for both root and subdomains

Whytecrowe committed:
update DomainToken tests to fix current ones and add new for the override logic

Whytecrowe committed:
remove some redundant isOwnerOf() and optimize registerSubdomain() AC and CoreRegistraArgs filling in RootRegistrar

Whytecrowe committed:
fix some simpler tests and helpers

Whytecrowe committed:
fix and extend RootRegistrar tests

Whytecrowe committed:
fix SubRegistrar test and some test helpers

Whytecrowe committed:
remove DOMAIN_TOKEN role and all logic related to it, including tests, helpers and deploy code

Whytecrowe committed:
Revert "remove DOMAIN_TOKEN role and all logic related to it, including tests, helpers and deploy code"

This reverts commit c97abf6eb5781af4f841f11ffe031cc77228b866.

Whytecrowe committed:
finish fixing existing tests

Whytecrowe committed:
rename new test file

JamesEarle committed:
resolve merge conflicts with stashed code for checkpoint

Whytecrowe committed:
update NatSpec with new changes

Whytecrowe committed:
fix SubRegistrarMock to be the same as the new main contract

Whytecrowe committed:
add some more checks for access control to controlled domains test

Whytecrowe committed:
extend current DomainToken transfer logic with more safety for burns and unsafe transfers, update NatSpec

Whytecrowe committed:
add tests to validate the new transfer safety logic

Whytecrowe committed:
small fixes

Whytecrowe committed:
fixes for linter

Whytecrowe committed:
fix type issues

Whytecrowe committed:
fix NatSpec for assignDomainToken()

Whytecrowe committed:
add missing test for isControlled()

Whytecrowe committed:
lint fix

MichaelKorchagin committed:
Review fixes. Needs tests. WIP

JamesEarle committed:
all tests passing for finalized pricer abstraction changes

JamesEarle committed:
cleanup, lint, remove comments and any debug code

JamesEarle committed:
extra cleanup

MichaelKorchagin committed:
Restoring test functionality.

MichaelKorchagin committed:
Mixed subdomains registration test. WIP

MichaelKorchagin committed:
Test to register mix of nested and non-nested subdomains.

MichaelKorchagin committed:
Linter fixes.

MichaelKorchagin committed:
Error name changed according to review.

JamesEarle committed:
fix after some feedback, all tests passing again

JamesEarle committed:
all tests, lint, and build

MichaelKorchagin committed:
Changed solution to two checks: hasRole method existence and admin role check. Corrected tests. WIP

JamesEarle committed:
add a few extra tests for length checks in validation and anything missing from coverage

JamesEarle committed:
add await for lint in CI

JamesEarle committed:
last few changes + dod addition

JamesEarle committed:
adjust workflow file structure

JamesEarle committed:
temp make always run for action

JamesEarle committed:
run on ubuntu-latest

JamesEarle committed:
edit yaml back to PR opened or edited

JamesEarle committed:
give path to file

JamesEarle committed:
remove testing for gh actions for dod, adjust rootPriceConfig tests to match small contract change

JamesEarle committed:
remove unnecessary address(0) check

MichaelKorchagin committed:
Added record checks to the register mix lvl subdomains test.

MichaelKorchagin committed:
Fixes according to Copilot comments.

Whytecrowe committed:
add ARegistrationPause abstract contract

Whytecrowe committed:
add pause logic to both registrars

Whytecrowe committed:
add tests

MichaelKorchagin committed:
Changed name of function. Added unchecked block in both: root and sub registrations. Added types for reg input data. Adjusted tests.

MichaelKorchagin committed:
Minor test fixes

MichaelKorchagin committed:
Test optimization

MichaelKorchagin committed:
Minor test fix

Whytecrowe committed:
simplify args var

MichaelKorchagin committed:
Merge branch 'rc/zchain-native-main' into mike1/cpt-67-bulk-domain-minting

Whytecrowe committed:
fix arg name in NatSpec

MichaelKorchagin committed:
Changed switch-case to if-statement in the test

Whytecrowe committed:
update solhint

Whytecrowe committed:
rename registration args in line with other arg structs and to prepare for the next merge

Whytecrowe committed:
fix natspec

JamesEarle committed:
fixes for last comments

Whytecrowe committed:
[v1.5] Bulk domain minting (#131)

JamesEarle committed:
lint

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v1.5/controlled-domains

Whytecrowe committed:
fix naming and adapt args structs and logic for the controlled domains feature

Whytecrowe committed:
fix type error

Whytecrowe committed:
[v1.5] Controlled Domains (#137)

JamesEarle committed:
Merge branch 'rc/zchain-native-main' into feat/polymorphic-pricer-configs, need to fix tests

JamesEarle committed:
remove unused error

JamesEarle committed:
all tests passing

JamesEarle committed:
Merge branch 'rc/zchain-native-main' into feat/polymorphic-pricer-configs

JamesEarle committed:
all tests passing, but there are type errors for some reason, seems from inconsistent dependency versions and the contract types generated from typechain

MichaelKorchagin committed:
Check CI commit. Added ADMIN_ROLE to the AAC and added IZNSRoles.

MichaelKorchagin committed:
Removed old compilers from HHConfig, updated proxy compilers from 0.8.20 to 22.

MichaelKorchagin committed:
Merge branch 'rc/zchain-native-main' into mike1/cpt-95-fix-a-bug-of-setting-wrong-accesscontroller

MichaelKorchagin committed:
Removed old proxy overrides in HHConfig

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v1.5/limited-access-registration-70

Whytecrowe committed:
add pause modifier to all the new registration functions on both registrars

Whytecrowe committed:
refactor some registration helpers to fund and approve user before registering both kinds of domains

Whytecrowe committed:
fix/move bulk tests and add new tests for pause when registering

Whytecrowe committed:
add solhint rules

Whytecrowe committed:
reorder vars in CoreRegisterArgs struct for more efficient packing

Whytecrowe committed:
fix linter errors

JamesEarle committed:
fix all build errors and tests, except 2 remaining in subregistrar

JamesEarle committed:
fix final test issues, clean, lint

JamesEarle committed:
add dangling comma

JamesEarle committed:
lint again

JamesEarle committed:
remove type and tokenOwner prop where not needed

MichaelKorchagin committed:
Changed try-catch block to make it catch errors correctly. Adjusted tests

Whytecrowe committed:
small fix

Whytecrowe committed:
fix NatSpec comment

Whytecrowe committed:
rename fixed price config to FixedPriceConfig

Whytecrowe committed:
pack DistributionConfig more efficiently

Whytecrowe committed:
small optimization in the pricer

Whytecrowe committed:
fix return type in the getter

Whytecrowe committed:
fix subregistrar mock to have the same order of props

Whytecrowe committed:
[v1.5] Polymorphic Price Config [WIP] (#135)

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v1.5/limited-access-registration-70

Whytecrowe committed:
fix all tests

Whytecrowe committed:
add public state var getters to registrar interfaces

Whytecrowe committed:
[v1.5] Limited Access to Registrations (#138)

Whytecrowe committed:
add missing test for #transferOverride()

Whytecrowe committed:
move upgrade mocks to the new test-fixtures folder

Whytecrowe committed:
update solcover config

MichaelKorchagin committed:
Added check for EOA. Review fixes, tests adjustment.

MichaelKorchagin committed:
Redundant function removal.

JamesEarle committed:
WIP beginnings of json creation for batch registration transaction proposal to safe on zchain

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into mike1/cpt-95-fix-a-bug-of-setting-wrong-accesscontroller

Whytecrowe committed:
linter fixes

Whytecrowe committed:
add missing import

Whytecrowe committed:
[v1.5] Fix a bug of setting wrong accesscontroller (#132)

Whytecrowe committed:
add test for assigning token to existing owner

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v1.5/cosmetic-improvements

JamesEarle committed:
registration and transfer scripts to create batches as JSON files we will upload to the safe

Whytecrowe committed:
fix linter script

Whytecrowe committed:
update natspec and small optimizations for AC and pricers

Whytecrowe committed:
update natspec and small optimizations for registrars

Whytecrowe committed:
update natspec and small optimizations for registry and resolvers

Whytecrowe committed:
update natspec and small optimizations for tokens, treasury and common errors

Whytecrowe committed:
update all docs files per contract

Whytecrowe committed:
fix for linter

Whytecrowe committed:
fix DomainToken.supportsInterface()

Whytecrowe committed:
update gas costs

Whytecrowe committed:
remove redundant comments

Whytecrowe committed:
rework price config validation for deploy campaign, root-registrar mission, fix tests and remove linter and compiler warnings

Whytecrowe committed:
fix error in getConfig()

Whytecrowe committed:
wrap all calls in deploy missions in the proper awaiter method from deployer

Whytecrowe committed:
update on-chain integration test and helpers, add configs for zephyr

Whytecrowe committed:
[v1.5] NatSpec update and optimizations (#142)

JamesEarle committed:
WIP snapshot script upgrades, some scrappy code while issues with zdc and sepolia verifications as a workaround, debugging tx failing in safe after execution

Whytecrowe committed:
rename "staking token address" vars to "root payment token address"

Whytecrowe committed:
add check for precisionMultiplier against a low price and return 0 when baseLength is zero without calculations

Whytecrowe committed:
update NatSpec

Whytecrowe committed:
remove redundant solhint disable

Whytecrowe committed:
linter fixes

JamesEarle committed:
push wip for kirill to copy

Whytecrowe committed:
rework logic to check for too large precision multiplier upon config validation instead of inside getPrice function. add test

Whytecrowe committed:
update NatSpec

JamesEarle committed:
updated api code, encoding working right with objects

JamesEarle committed:
Merge branch 'rc/zchain-native-main' into v1.5/registration-script-92

JamesEarle committed:
wip push update for kirill

JamesEarle committed:
working, add notes and restructure for cleanliness. Comments detailing usage

JamesEarle committed:
add loop around subdomains to iterate all depths individually

JamesEarle committed:
small updates to other scripts to match updates in main

Whytecrowe committed:
make postDeploy task to pause RootRegistrar if needed

Whytecrowe committed:
Revert "make postDeploy task to pause RootRegistrar if needed"

This reverts commit b5e9bc056d968168ee1ad7dba893e2bf2637d157.

Whytecrowe committed:
Reapply "make postDeploy task to pause RootRegistrar if needed"

This reverts commit 3e0b75128b9e664b0a072937d8a2c2488d5748ba.

Whytecrowe committed:
add registration pause post deploy to campaign

- add PAUSE_REGISTRATION env var
- pass it through the campaign config
- add postDeploy task to both registrars that pauses registration if env var is set
- add test

Whytecrowe committed:
fix wrong bool return from needsPostDeploy on registrars

Whytecrowe committed:
add V1.1. audit report

Whytecrowe committed:
rename audit report file

Whytecrowe committed:
rename dir

Whytecrowe committed:
add rootPaymentType state var to determine payments for all root domains

Whytecrowe committed:
add env var to determine rootPaymentType and pass that through campaign config

Whytecrowe committed:
update RootRegistrar DM to get rootPaymentType from the config and deploy properly

Whytecrowe committed:
add new tests for deploy and root registrations, update existing tests and helpers

JamesEarle committed:
WIP still, some unminted lvl1 subs that cause lvl2 reverts, not in db somehow, refreshing data

JamesEarle committed:
a few extras before refreshing data in v1.0 branch

JamesEarle committed:
pr changes and recreation + proposal of missing parent domains instead of just storing in DB.

JamesEarle committed:
linting

JamesEarle committed:
more linting

JamesEarle committed:
Merge branch 'rc/zchain-native-main' into v1.5/registration-script-92

JamesEarle committed:
small fix to incorrept change, updated types

JamesEarle committed:
resolve build issues

JamesEarle committed:
added write logs to db, as well as additional comments on env vars

JamesEarle committed:
comment out unneeded hh config settings for CI

JamesEarle committed:
more comments out for CI

JamesEarle committed:
address linter issues in CI

JamesEarle committed:
comment out networks

Whytecrowe committed:
Add scripts that create bulk register and transfer domain functions through the Safe API (#143)

Whytecrowe committed:
small fixes to unblock potential stuck txes in Safe

JamesEarle committed:
Decouple registration batch and transfer batch creation for improved separation downstream in later executions

JamesEarle committed:
Fix - Incorrect array used when registering missing revoked parent domains

JamesEarle committed:
Fix - Incorrect domain token recipient causes ownership discrepancy after migration

Whytecrowe committed:
[Low] Unbounded recursion in dropDB can exhaust system resources

JamesEarle committed:
Fix - Duplicate revoked parent causes full batch revert during registration

JamesEarle committed:
address minor linter warnings

JamesEarle committed:
Fix - Incorrect parent and parentHash assigned to missing parents

JamesEarle committed:
Fix - Inconsistent priority order for parent hash resolution in subdomain registration

JamesEarle committed:
Fix - Original domain address is not preserved in V2 contract registration

Whytecrowe committed:
[Info] Protocol fee can Be deducted from Staked funds

Whytecrowe committed:
[Info] Missing minimum price enforcement in _getPrice

MichaelKorchagin committed:
Audit fix. Changed admin check to governor when #setAC

JamesEarle committed:
Fix - A Single token owner can block all transfers in batch

Whytecrowe committed:
ZNS v2.0 Audit Responses (#145)

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v2/audit-fixes-1

JamesEarle committed:
address pr feedback

JamesEarle committed:
modify to run estimation on every transfer

MichaelKorchagin committed:
Added "revoke" action to execute revoke batch. Added a function to create revoke batches as well.

Whytecrowe committed:
remove redundant extra deploy in test

MichaelKorchagin committed:
Initial test. Sends tx batches. Some corrections of the script.

JamesEarle committed:
Merge branch 'rc/zchain-native-main' into native-main/audit-fix-1

Whytecrowe committed:
Fix - A Single token owner can block all transfers in batch (#150)

Whytecrowe committed:
small fixes

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into mike1/cpt-149-audit-fix-setting-wrong-access-controller-bug

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into mike1/cpt-149-audit-fix-setting-wrong-access-controller-bug

Whytecrowe committed:
Audit fix / "Admin can replace the Governor role" (#149)

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into v2/audit-fixes-1

Whytecrowe committed:
[V2] Audit Fixes (Kirill) (#147)

Whytecrowe committed:
Merge branch 'rc/zchain-native-main' into mike/feat-158/revoke-batches

MichaelKorchagin committed:
Review fixes. Execution's transferred. Test saturation.

MichaelKorchagin committed:
Merge branch 'mike/feat-158/revoke-batches' of https://github.com/zer0-os/zNS into mike/feat-158/revoke-batches

JamesEarle committed:
domain migration test passing

Whytecrowe committed:
add logic to NOT override env vars from .env file if we are not specifying the network/running on local hardhat network

Whytecrowe committed:
remove checks for hardhat network - it will not run there, simplify logger call in execution

Whytecrowe committed:
fix comments

Whytecrowe committed:
[AUDIT] Revoked domains remain locked under Safe ownership post migration (#152)

Whytecrowe committed:
improve HH config to dynamically get the env vars and default values

Whytecrowe committed:
move env setup to the top of the HH config

Whytecrowe committed:
improve environment logic

Whytecrowe committed:
add price config encoder script

Whytecrowe committed:
small fixes for registration flow

Whytecrowe committed:
uncomment error logging for tx estimations

Whytecrowe committed:
add commented out code to turn on whenever domains have not been registered

Whytecrowe committed:
add revoking and polish

Whytecrowe committed:
add logging for revokes and remove safeKit gas estimation

Whytecrowe committed:
silence eslint rule

Whytecrowe committed:
fix compiler error in test

Whytecrowe committed:
revert test code for transfers back to the prod code

Whytecrowe committed:
update documentation

Whytecrowe committed:
add latest audit report

Whytecrowe committed:
V2 Doc Updates (#156)
